src/Proto3/Suite/DotProto/Generate.hs

```diff
@@ -826,48 +834,103 @@ toJSONPBMessageInstD _ctxt parentIdent msgIdent messageParts = do
               , HsVar (unqual_ varNm)
               ]
 
-  -- E.g.
-  -- case f4_or_f9 of
-  --   Just (SomethingPickOneName f4)
-  --     -> HsJSONPB.pair "name" f4
-  --   Just (SomethingPickOneSomeid f9)
-  --     -> HsJSONPB.pair "someid" f9
-  --   Nothing
-  --     -> mempty
-  let oneofCaseE (OneofField{subfields}) =
-        HsCase disjunctName (altEs <> pure fallThruE)
+  -- Note: [Generating eta-reduced code]
+  -- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  -- Suppose we have a sum type Foo, nested inside a message Bar.
+  -- We want to generate the following:
+  --
+  -- > toJSONPB (Bar foo more stuff) =
+  -- >   HsJSONPB.object
+  -- >     [ (\caseExpr option -> if optEmitInlinedOneof option
+  -- >                            then oneofCaseE option
+  -- >                            else PB.object ["Foo" .= (PB.object [caseExpr] option)] option
+  -- >       ) (<case expr scrutinising foo> :: Options -> Value)
+  -- >     , <do more>
+  -- >     , <do stuff>
+  -- >     ]
+  --
+  -- To avoid gensymming parameters for the lambda, we will instead generate the
```

Comment:
I don't think we should use the reader monad just to avoid generating unique names.  This makes the generated code hard for me to read.  For example `((flip ("Foo" .=) <*>) . PB.object . pure)` is difficult for me to read